### PR TITLE
dashboard location reassignment

### PR DESCRIPTION
**Pull request by calellowitz**
##### SUMMARY

This is the aggregation work for the new location reassignment. I expect to play with how the queries are invoked (to make them more parallel), but do not expect to change the queries themselves. I also need to add some more tests. Currently the UCR part is tested via the person UCR, but the intermediate tables and the infra part need tests. Leaving this with the do not merge tag for those reasons, but wanted to start with feedback on the approach.

I did benchmark some of these queries and it added 10-20 minutes, which is not ideal but manageable. I am going to do some more benchmarking now that there is time after the agg runs.

Best reviewed 🐡 

##### TECHNICAL OVERVIEW
The basic premise of this code is that starting soon, we need to support routine changes to the location hierarchy. This mostly manifests in the dashboard as cases being reassigned to new owners. The fundamental goal of this PR is to change the aggregation such that if a case is moved, it is aggregated as part of its previous location before the change, and its new location after. Since the dashboard only aggregates two months of data, that means we need to support a situation where the case has a different owner in each of the two months.

One unique challenge of these requirements is that citus requires the shard key to be included in joins for distributed tables, and location movement can change the shard key for a case, since we shard by supervisor.

The way this PR deals with that is to create copies of the relevant case UCRs for previous months with location info as of that time. These are used to also create copies of some intermediate tables since we need join those to prior month versions themselves. The copied tables are then properly distributed for all other queries, and can be a drop in replacement for the regular UCR or intermediate table.

Creating those tables was a bit trickier than expected since citus does not let you update the shard column, which is why I used `DELETE` followed by `INSERT` and did the updates locally on the master node.

The general spec for location reassignment work is https://docs.google.com/document/d/1h6kTr9M0FumswP3-gr_84OLeUHRmn5Y10dCpSAS3Y9g/edit#

Dashboard specific information is also in https://docs.google.com/document/d/1LO9oO3Z-JQbeTMr-76sgyVDMFfYWM5eItEvTxJq2E2Y/edit

### CODE REVIEW

**Comment by calellowitz** [comment]
@mkangia 
I am aware that I need to change the names in the UCR columns, and will get to that soon, but wanted this up first.

File: custom/icds_reports/ucr/data_sources/ccs_record_cases.json
Comment (czue) [inline custom/icds_reports/ucr/data_sources/ccs_record_cases.json:1252]: is there a better phrase we can come up with to `location_rationalisation`? where did that come from? those words just don't fit together in any meaningful way in my brain and i imagine that will be likely to be true for anyone who visits this code later

File: custom/icds_reports/models/aggregate.py
Comment (czue) [inline custom/icds_reports/models/aggregate.py:243]: assume it's intentional that these are not `datetime`s?

File: custom/icds_reports/utils/aggregation_helpers/distributed/agg_awc.py
Comment (czue) [inline custom/icds_reports/utils/aggregation_helpers/distributed/agg_awc.py:187]: :snail:

File: custom/icds_reports/utils/aggregation_helpers/distributed/agg_awc.py
Comment (czue) [inline custom/icds_reports/utils/aggregation_helpers/distributed/agg_awc.py:454]: what's this for?

File: custom/icds_reports/utils/aggregation_helpers/distributed/agg_awc.py
Comment (czue) [inline custom/icds_reports/utils/aggregation_helpers/distributed/agg_awc.py:41]: what's the human-readable explanation of this check?

File: custom/icds_reports/utils/aggregation_helpers/distributed/ccs_record_monthly.py
Comment (czue) [inline custom/icds_reports/utils/aggregation_helpers/distributed/ccs_record_monthly.py:38]: now seeing this repeated makes me think it should be a well-named helper function

File: custom/icds_reports/utils/aggregation_helpers/distributed/deprecation.py
Comment (czue) [inline custom/icds_reports/utils/aggregation_helpers/distributed/deprecation.py:46]: nit: should choose between british and american spelling if we're sticking with this terminology

File: custom/icds_reports/utils/aggregation_helpers/distributed/deprecation.py
Comment (czue) [inline custom/icds_reports/utils/aggregation_helpers/distributed/deprecation.py:1]: `deprecation.py` is a bit of a funny module name just because of the Python concept. maybe fine, just pointing it out.

**Review by czue — COMMENTED** [review]
I did a cursory review because I don't really have my head wrapped around what's going on. Is there a spec / design doc that explains what this is actually doing at a high level?

File: custom/icds_reports/ucr/data_sources/ccs_record_cases.json
Comment (calellowitz) [reply custom/icds_reports/ucr/data_sources/ccs_record_cases.json:1252]: The phrase was chosen because it was the original name of this project and to be consistent with the case properties defined in https://docs.google.com/document/d/1h6kTr9M0FumswP3-gr_84OLeUHRmn5Y10dCpSAS3Y9g/edit#. It has since been renamed `location_reassignment`, and all of these will be updated to that (that change was quite recent. The change to the case property names were made in the last few days which is why it is not yet reflected

File: custom/icds_reports/models/aggregate.py
Comment (calellowitz) [reply custom/icds_reports/models/aggregate.py:243]: Yeah, time does not matter for these and I thought that dates would make comparisons easier

File: custom/icds_reports/utils/aggregation_helpers/distributed/agg_awc.py
Comment (calellowitz) [reply custom/icds_reports/utils/aggregation_helpers/distributed/agg_awc.py:41]: "Is this the current month agg?" is the human readable version but will pull it into a better named function

File: custom/icds_reports/utils/aggregation_helpers/distributed/deprecation.py
Comment (calellowitz) [reply custom/icds_reports/utils/aggregation_helpers/distributed/deprecation.py:46]: These will all be reassignment soon, but yep. They were all american until I realized the spec called for the case property to use the british spelling. Fixed all those cases but must have missed this

File: custom/icds_reports/utils/aggregation_helpers/distributed/deprecation.py
Comment (calellowitz) [reply custom/icds_reports/utils/aggregation_helpers/distributed/deprecation.py:1]: I think I may get rid of this entirely, and try to integrate this functionality into the existing agg helpers for the relevant steps, but if it sticks around, I can rename it. How does "location_reassignment" sound? That indicates it's functionality.

**Comment by calellowitz** [comment]
@czue updated the PR description to have a much longer overview of the ideas here, and added links to the specs. Let me know if you still have questions.

File: custom/icds_reports/utils/aggregation_helpers/distributed/deprecation.py
Comment (czue) [reply custom/icds_reports/utils/aggregation_helpers/distributed/deprecation.py:1]: "location_reassignment" sounds good to me :+1:

File: custom/icds_reports/models/aggregate.py
Comment (mkangia) [reply custom/icds_reports/models/aggregate.py:243]: just thinking if these should be kept as `datetime` as well so mirror what HQ is keeping. I understand the time is kind of irrelevant for dashboard but if it does not hurt in terms of performance etc why not keep it the same as other places?

Comment (calellowitz) [reply custom/icds_reports/models/aggregate.py:243]: The dashboard only cares about days, and in this case, only cares about months (date with day of 1, so 2020-03-01). That means this will only be compared to dates and we want it to always be treated as the date that it represents, or a month truncated version, not a datetime. I think leaving it as a datetime in the database opens us up to a number of comparison bugs (`2020-03-01 01:04 > 2020-03-01`) that could otherwise be avoided by making sure everything is a date to begin with, rather than having to remember to coerce it each time

Comment (mkangia) [reply custom/icds_reports/models/aggregate.py:243]: cool. thanks for explaining this @calellowitz

**Comment by czue** [comment]
Thanks for adding the additional info @calellowitz . I'm still not sure I'm best positioned to review this in detail due to only surface level understanding of the agg, but I'm happy to take a second pass once you've updated.

**Comment by calellowitz** [comment]
@czue 
Updated. I am not totally sure who is best placed to review this, so would love if you would take another pass, but if you have other suggestions happy to reach out to them as well.

File: custom/icds_reports/utils/aggregation_helpers/distributed/awc_infrastructure.py
Comment (czue) [inline custom/icds_reports/utils/aggregation_helpers/distributed/awc_infrastructure.py:27]: And either way is there a name that might be more descriptive that could be used? e.g. `get_table_name_for_previous_months_agg` (not 100% sure that's right)

File: custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py
Comment (czue) [inline custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py:34]: where is `CREATE_QUERY` defined?

File: custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py
Comment (czue) [inline custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py:10]: nit: maybe change to `TempPrevTablesBase` so it's clearer that this class is only meant to be used as a base class?

bonus: use `abc` to make it an actual abstract base classs

File: custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py
Comment (czue) [reply custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py:34]: nvm, i get it

File: custom/icds_reports/utils/aggregation_helpers/distributed/agg_awc.py
Comment (czue) [reply custom/icds_reports/utils/aggregation_helpers/distributed/agg_awc.py:197]: I think this is the answer: https://github.com/dimagi/commcare-hq/pull/27083/files#diff-0231c9efff8817406d8b2185226a540eR48

File: custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py
Comment (czue) [inline custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py:98]: Are these four lines this bit? Having a hard time wrapping my head around them.

> Creating those tables was a bit trickier than expected since citus does not let you update the shard column, which is why I used DELETE followed by INSERT and did the updates locally on the master node.

File: custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py
Comment (czue) [inline custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py:109]: would a better name for the key be `ucr_prev_local`?

File: custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py
Comment (czue) [inline custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py:116]: it seems like it could be easy to miss updating this list if we add a new table. is there test coverage that would catch that? or is there any way the code could be restructured such that it'd be more intuitive / natural that it would get updated? maybe even moving the list into some constants file might help...

File: custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py
Comment (czue) [inline custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py:136]: what are your thoughts on maybe extracting this to a file (or at least adding some newlines)? this line in particular seems to have a lot going on and is quite difficult to read.

File: custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py
Comment (czue) [inline custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py:75]: same comment as below re: the table list

**Review by czue — COMMENTED** [review]
Alright I tried to do my best given the context I had. Holding off on approval till you respond to comments and also I think @Rohit25negi should review `location_reassignment.py` in detail as I don't feel like I understand the flow of UCR / intermediate / infrastructure tables well enough to validate that the SQL there is correct. But big picture I don't think any of my feedback is blocking.

Also I didn't review any of the csv and sql diff files. let me know if that seems like a problem.

File: custom/icds_reports/ucr/data_sources/ccs_record_cases.json
Comment (calellowitz) [reply custom/icds_reports/ucr/data_sources/ccs_record_cases.json:1261]: I thought the case property name was kind of unwieldy, and this would be a little clearer. Plus they look kind of similar if you leave the whole thing since a multiple would have the form `location_reassignment_last_X_id`. I think the differences are a little clearer this way.

File: custom/icds_reports/utils/aggregation_helpers/distributed/agg_awc.py
Comment (calellowitz) [reply custom/icds_reports/utils/aggregation_helpers/distributed/agg_awc.py:203]: UCR table

File: custom/icds_reports/utils/aggregation_helpers/distributed/agg_awc.py
Comment (calellowitz) [reply custom/icds_reports/utils/aggregation_helpers/distributed/agg_awc.py:197]: Yeah, it allowed consistency across all the UCRs, and I think the name is slightly clearer too

File: custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py
Comment (calellowitz) [reply custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py:98]: Yeah, there is a similar version in all three of these classes, but this version is, in order:
1. creates a local copy of all the rows in the previous month's table that need to be updated (were migrated at the start of the next month).
2. delete those same rows from the previous month's table
3. modifies the local copy created in step 1 to update location info to previous owners
4. inserts the local copy data into the previous month table

File: custom/icds_reports/utils/aggregation_helpers/distributed/awc_infrastructure.py
Comment (calellowitz) [reply custom/icds_reports/utils/aggregation_helpers/distributed/awc_infrastructure.py:27]: The UCR one is a complete copy of the data, slightly modified. The `BaseICDSAggregationDistributedHelper` is using a monthly table, and is a copy of the previous month's data, hence `prev`. I think that one could be changed to `temp_prev_tablename` but the previous part is much more meaningful on these (its old data), vs the UCRs where it is a modified copy.

Comment (calellowitz) [reply custom/icds_reports/utils/aggregation_helpers/distributed/awc_infrastructure.py:27]: maybe `get_ucr_tablename_from_previous_months_agg` but yeah. i will think on some of this naming stuff overnight and try to improve these

File: custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py
Comment (calellowitz) [reply custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py]: Added some formatting to make this clearer, let me know if you still prefer another file

File: custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py
Comment (calellowitz) [reply custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py]: Added a test for the intermediate tables one. Not sure how to test the UCR ones, but can pull into a set of constants if thats easier. The general idea is that its any case UCR

**Comment by calellowitz** [comment]
@czue @mkangia I think this is ready for review again

File: custom/icds_reports/utils/aggregation_helpers/distributed/awc_infrastructure.py
Comment (czue) [reply custom/icds_reports/utils/aggregation_helpers/distributed/awc_infrastructure.py:27]: got it thanks.

File: custom/icds_reports/tests/agg_tests/__init__.py
Comment (czue) [inline custom/icds_reports/tests/agg_tests/__init__.py:37]: maybe is fine, but does this mean we will no longer have test coverage for our production agg until this gets rolled out?

File: custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py
Comment (czue) [inline custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py:146]: was this meant to be here?

**Review by czue — COMMENTED** [review]
Reviewed the last commits and everything looks ok except for a potentially blocking typo.

Might do one final larger pass if I have time but wanted to get this out in case :baby: wakes up soon

One bigger picture question - I know you've done some performance testing already but are you quite confident the increased DB size won't cause any issues? I don't think I realized until just now that were duplicating entire UCR tables.

**Comment by czue** [comment]
actually one request - maybe you can resolve all comments you've addressed?

**Comment by calellowitz** [comment]
@czue Just resolved all the ones that were clearly addressed. I left a few open ones that I think require your judgement whether they now meet your (or whether your question was answered).

**Review by Rohit25negi — CHANGES_REQUESTED** [review]
hey, Left a couple of comments.

File: custom/icds_reports/utils/aggregation_helpers/__init__.py
Comment (Rohit25negi) [inline custom/icds_reports/utils/aggregation_helpers/__init__.py:37]: Are we sure about the robustness of this during the month start. Like if the agg is kicked of on 31st night. and when it gets to agg_Awc it is already next month. What is the behavior we expect then?

File: custom/icds_reports/utils/aggregation_helpers/distributed/awc_location.py
Comment (Rohit25negi) [inline custom/icds_reports/utils/aggregation_helpers/distributed/awc_location.py:82]: why the variable names are ls_deprecates not supervisor_deprecates which matches to the existing naming?

File: custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py
Comment (Rohit25negi) [inline custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py:81]: are we intentionally only include dashboard case ucrs?

File: custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py
Comment (Rohit25negi) [inline custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py:43]: should it also include `INCLUDE CONSTRAINS INDEXES`?

File: custom/icds_reports/ucr/data_sources/ccs_record_cases.json
Comment (Rohit25negi) [inline custom/icds_reports/ucr/data_sources/ccs_record_cases.json:1252]: I assume all these fields will have data going forward so no need to backfill yeah?

**Review by Rohit25negi — CHANGES_REQUESTED** [review]
some comments.

File: custom/icds_reports/ucr/data_sources/ccs_record_cases.json
Comment (calellowitz) [reply custom/icds_reports/ucr/data_sources/ccs_record_cases.json:1252]: exactly

File: custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py
Comment (calellowitz) [reply custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py:81]: yeah, form UCRs do not have this same problem because the forms are already associated with the correct location for a given month. so a submission in march will have march location but an april submission will be inserted with the april location. Cases only have the latest info, so once it is updated in April it will no longer contain march info

File: custom/icds_reports/utils/aggregation_helpers/__init__.py
Comment (calellowitz) [reply custom/icds_reports/utils/aggregation_helpers/__init__.py:37]: That is a great question. I think I may have to rethink this one

File: custom/icds_reports/utils/aggregation_helpers/distributed/awc_location.py
Comment (calellowitz) [reply custom/icds_reports/utils/aggregation_helpers/distributed/awc_location.py:82]: No reason. I can change that

File: custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py
Comment (calellowitz) [reply custom/icds_reports/utils/aggregation_helpers/distributed/location_reassignment.py:43]: I did not include those on purpose, since they slow down the insert. I will add ones that are used by the agg, but do in separately at the end so that insert is fast and we don't have unnecessary indices

File: custom/icds_reports/tests/agg_tests/__init__.py
Comment (calellowitz) [reply custom/icds_reports/tests/agg_tests/__init__.py:37]: Yeah, it sort of a pick your poison situation, either the new agg won't have tests or the old one won't, but the risk that new changes break this logic seems higher than the old logic, and beyond the source tables, they are basically identical processes, so new changes shouldn't exercise the differences anyway.

**Comment by calellowitz** [comment]
> One bigger picture question - I know you've done some performance testing already but are you quite confident the increased DB size won't cause any issues? I don't think I realized until just now that were duplicating entire UCR tables.

I did think about this a bit. The largest table this duplicates is still single digit GB per node. And we only ever have one copy at a time. This is a fixed disk cost (one copy of each of these), rather than one that grows each month.

File: custom/icds_reports/utils/aggregation_helpers/__init__.py
Comment (calellowitz) [reply custom/icds_reports/utils/aggregation_helpers/__init__.py:37]: On second pass, I don't actually think there is a downside to this. Worst case it uses the temp tables, rather than the live UCRs, but those will exist and contain correct data anyway, so that sounds fine.

**Comment by calellowitz** [comment]
@czue and @Rohit25negi 
This is ready for a (hopefully) final review. I think I answered all questions and looking to get this out this week for QA

File: custom/icds_reports/tasks.py
Comment (czue) [inline custom/icds_reports/tasks.py:1850]: are these supposed to be `helper.helper`?

Comment (czue) [reply custom/icds_reports/tasks.py:1850]: also not sure if worth it, but if each of these is always going to do the same thing you could pull these two-liners into a function and then these all become one-liners. (and if there are changes to what they do they only have to get edited in that one function)

**Review by czue — COMMENTED** [review]
just reviewed the last couple commits. build fail looks real

**Review by Rohit25negi — APPROVED** [review]
looks good to me

**Review by czue — APPROVED** [review]
I didn't re-review but I've been through this enough times that I figured I should give you a stamp.

**Comment by Rohit25negi** [comment]
@calellowitz  Curious to know your thoughts on this. But It feels like this would affect the way we perform the data pulls for multiple months combines in one or similar data pulls.  How do you think it would affect that.

Also, will it affect the historical data corrections we do.?

**Pull request merged**
